crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1084,12 +996,212 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
+                                ops.push(Operation::Remove(i - 1));
+                            } else {
+                                trace!("replacing day divider @ {i} with new timestamp from event");
+                                ops.push(Operation::Replace(i - 1, ts));
+                            }
+                        }
+                    } else if let Some(prev_event) = prev_item.as_event() {
+                        // The event is preceded by another event. If they're not the same date,
+                        // insert a date divider.
+                        let prev_ts = prev_event.timestamp();
+                        if timestamp_to_date(prev_ts) != timestamp_to_date(ts) {
+                            trace!("inserting day divider @ {} between two events with different dates", i);
+                            ops.push(Operation::Insert(i, ts));
+                        }
+                    }
+                } else {
+                    // The event was the first item, so there wasn't any day divider before it:
+                    // insert one.
+                    trace!("inserting the first day divider @ {}", i);
+                    ops.push(Operation::Insert(i, ts));
+                }
+
+                prev_item = Some((item, ts));
+                last_event_ts = Some(ts);
+            }
+        }
+
+        // Record the deletion offset.
+        let mut offset = 0;
+        // Remember what the maximum index was, so we can assert that it's non-decreasing.
+        let mut max_i = 0;
+
+        for op in ops {
+            match op {
+                Operation::Insert(i, ts) => {
+                    assert!(i >= max_i);
+                    self.items.insert(
+                        i + offset,
+                        self.meta.new_timeline_item(VirtualTimelineItem::DayDivider(ts)),
+                    );
+                    offset = (offset + 1).min(self.items.len());
+                    max_i = i;
+                }
+
+                Operation::Replace(i, ts) => {
+                    assert!(i >= max_i);
+                    let prev = &self.items[i + offset];
+                    assert!(prev.is_day_divider(), "we replaced a non day-divider");
+
+                    let unique_id = prev.unique_id();
+                    let item = TimelineItem::new(VirtualTimelineItem::DayDivider(ts), unique_id);
+
+                    self.items.set(i + offset, item);
+                    max_i = i;
+                }
+
+                Operation::Remove(i) => {
+                    assert!(i >= max_i);
+                    let prev = self.items.remove(i + offset);
+                    assert!(prev.is_day_divider(), "we removed a non day-divider");
+                    offset = offset.saturating_sub(1);
+                    max_i = i;
+                }
+            }
+        }
+
+        // Then check invariants.
+        self.assert_date_divider_invariants();
```

Comment:
I love this approach. Should it return bool and then we `assert!` here? Then nothing would run if asserts are disabled.

Reply:
(Yep, modified in a subsequent commit!)